keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,487 @@
+for Job queueing or workflows to support all of them.
+
+[Job patterns]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#job-patterns
+
+### Goals
+
+- Support the *array Job* pattern by adding completion indexes to each Pod
+  of a Job in *fixed completion count* mode.
+
+### Non-Goals
+
+- Support for work lists, where each Pod receives a different element of a
+  static list. This can be implemented by users from completion indexes.
+- Support for completion index in non-parallel Jobs or Jobs with a work queue.
+- All-or-nothing scheduling.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Job author, I can create an array Job where each Pod receives a unique and
+ordered completion index. I can use the index in my binary through an
+environment variable or a file.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: parallel-work
+spec:
+  completions: 100
+  parallelism: 100
+  template:
+    spec:
+      containers:
+      - name: task
+        image: registry.example.com/processing-image
+        command: ["./process",  "--index", "$INDEX"]
+        env:
+        - name: INDEX
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.annotations['alpha.batch.k8s.io/job-completion-index'] 
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+* An earlier proposal for [indexed Job]
+suggested the support for work lists, i.e. passing different parameters to each
+Pod. We decided to leave this out of the proposal to keep it simple and
+because work lists can be implemented in a startup script using the completion
+index as building block.
+* The semantics of an indexed Job are similar to a StatefulSet, in the sense
+that Pods have an associated index and Pods.
+However, the APIs have a major difference: a StatefulSet doesn't have completion
+semantics, as opposed to Jobs.
+
+[indexed Job]: https://github.com/kubernetes/community/blob/b21d1b27c8c748bf81283c2d89cde2becb5f2709/contributors/design-proposals/apps/indexed-job.md
+
+### Risks and Mitigations
+
+Jobs have a known issue in which more than one Pod can be started even if
+parallelism and completion are set to 1 ([reference]). In the case of indexed
+Jobs, this translates as more than one Pod having the same index.
+
+Just like for existing Job patterns, workloads have to handle duplicates at the
+application level.
+
+[reference]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures
+
+## Design Details
+
+### Indexed Jobs
+
+Indexed Jobs are Jobs [with fixed completion count], that is, Jobs that have a
+non-nil `.spec.completions` field.
+
+### Pod detail
+
+The Pod and PodSpec APIs don't get any new fields. However, Pods created for
+Indexed Jobs get the annotation `alpha.batch.k8s.io/job-completion-index` with a
+value equal to its completion index. The annotation is immutable.
+
+The annotation can be accessed through the downward API as a file or environment
```

Comment:
would that require a kubelet change? if so, perhaps it is not worth the dependency?

Reply:
I also prefer we use the existing downward API, rather than silently adding the variable. Note that the same can be achieved with a mutating webhook too.

Note that we still require having the index somewhere else (like annotations) so that the controller can use it for tracking.

Reply:
Not a kubelet change.  I am suggesting the following logic in Job controller (pseudocode):
```
// Make new Pod
pod = new Pod
// Start with user-provided values
pod.mergeValues(job.spec.template.spec)
// set a default value for the env var, unless there is a collision.
for container in pod.spec.containers:
    if pod.spec.containers.env[] does not contain item with name = "INDEX":
        ev =  new ContainerEnvVar
        ev.name = "INDEX"
        ev.valueFrom.fieldRef.fieldPath =  "metadata.annotations['alpha.batch.k8s.io/job-completion-index']"
        pod.spec.containers.env.append(ev)
create(pod)
```

Reply:
I put this for alpha